jax/_src/custom_call.py

```diff
@@ -0,0 +1,117 @@
+# Copyright 2024 The JAX Authors.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     https://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+from __future__ import annotations
+
+from collections.abc import Iterable, Sequence
+from typing import Any
+
+import numpy as np
+
+from jax._src import core
+from jax._src import dtypes
+from jax._src.interpreters import mlir
+from jax._src.interpreters.mlir import ir
+
+
+def custom_call_lowering(
+  call_target_name: str,
+  api_version: int = 4,
```

Comment:
I wonder if we need to expose `api_version=` here at all. It looks like the new typed FFI is objectively better and more user friendly than its predecessors, so maybe we should just hardcode `api_version=4` and require `backend_config` to be a `dict`/`Mappping`?

Reply:
Otherwise, it is easy for the caller to get unsupported combination of arguments, e.g. 

```
custom_call_lowering(api_version=4, backend_config="foo")
```

Reply:
I think agree here with the caveat that maybe we want to support specifying `api_version >= 4` for future compatibility? When the next API version comes out, I'm not sure we'd want a hard coded version here... but I'm not convinced that I know that best way to handle this!

Reply:
Yeah, I don't have a strong opinion here, but maybe @hawkinsp does?

Reply:
I think I'm pretty happy just setting the default for `api_version` to `4`. Given the API contract for `extend`, I don't think it's the end of the world to cross this bridge if we get there.